linera-execution/src/execution.rs

```diff
@@ -30,20 +32,37 @@ use {
 use super::{execution_state_actor::ExecutionRequest, runtime::ServiceRuntimeRequest};
 use crate::{
     execution_state_actor::ExecutionStateActor, resources::ResourceController,
-    system::SystemExecutionStateView, ApplicationDescription, ApplicationId, ExecutionError,
-    ExecutionRuntimeConfig, ExecutionRuntimeContext, MessageContext, OperationContext,
-    ProcessStreamsContext, Query, QueryContext, QueryOutcome, ServiceSyncRuntime, Timestamp,
-    TransactionTracker,
+    system::SystemExecutionStateView, ApplicationDescription, ApplicationId, BcsHashable,
+    Deserialize, ExecutionError, ExecutionRuntimeConfig, ExecutionRuntimeContext, MessageContext,
+    OperationContext, ProcessStreamsContext, Query, QueryContext, QueryOutcome, Serialize,
+    ServiceSyncRuntime, Timestamp, TransactionTracker,
 };
 
 /// A view accessing the execution state of a chain.
-#[derive(Debug, ClonableView, CryptoHashView, Allocative)]
+#[derive(Debug, ClonableView, View, Allocative)]
 #[allocative(bound = "C")]
 pub struct ExecutionStateView<C> {
     /// System application.
-    pub system: SystemExecutionStateView<C>,
+    pub system: HistoricallyHashableView<C, SystemExecutionStateView<C>>,
     /// User applications.
-    pub users: HashedReentrantCollectionView<C, ApplicationId, KeyValueStoreView<C>>,
+    pub users: HistoricallyHashedReentrantCollectionView<C, ApplicationId, KeyValueStoreView<C>>,
+}
+
+impl<C> ExecutionStateView<C>
+where
+    C: Context + Clone + Send + Sync + 'static,
+    C::Extra: ExecutionRuntimeContext,
+{
+    pub async fn crypto_hash_mut(&mut self) -> Result<CryptoHash, ViewError> {
+        #[derive(Serialize, Deserialize)]
+        struct ExecutionStateViewHash([u8; 32]);
+        impl BcsHashable<'_> for ExecutionStateViewHash {}
+        let mut hasher = linera_views::sha3::Sha3_256::default();
+        hasher.update_with_bytes(&self.system.historical_hash().await?)?;
+        hasher.update_with_bytes(&self.users.historical_hash().await?)?;
```

Comment:
Actually, can we `let ExecutionStateView { … } = self`, so the compiler would complain if we added a field, and we can't forget it here?